File: sdk/python/src/bsdkrun/process.py
"""Run the ``bsdkrun`` CLI and capture its output.

Every invocation is prepended with the global ``--log-level`` flag (default 0)
so the SDK's captured output stays clean. Raise it for boot diagnostics.
"""

from __future__ import annotations

import os
import subprocess
import threading
from collections.abc import Callable, Mapping
from dataclasses import dataclass

from .binary import resolve_binary
from .errors import CommandFailed

__all__ = ["BinaryResult", "RawResult", "run", "run_binary", "run_checked", "spawn"]


@dataclass(frozen=True)
class RawResult:
    """The buffered result of a ``bsdkrun`` invocation."""

    stdout: str
    stderr: str
    exit_code: int


@dataclass(frozen=True)
class BinaryResult:
    """Like :class:`RawResult`, but stdout is left as bytes."""

    stdout: bytes
    stderr: str
    exit_code: int


def _with_globals(args: list[str], log_level: int | None) -> list[str]:
    """Prepend the global ``--log-level`` flag."""
    return ["--log-level", str(log_level if log_level is not None else 0), *args]


def _run_raw(
    args: list[str],
    *,
    env: Mapping[str, str] | None = None,
    stdin: str | bytes | None = None,
    log_level: int | None = None,
    on_stdout: Callable[[bytes], None] | None = None,
    on_stderr: Callable[[bytes], None] | None = None,
) -> tuple[bytes, bytes, int]:
    """Run ``bsdkrun <args>`` and return ``(stdout, stderr, exit_code)`` as bytes.

    The undecoded form both :func:`run` and :func:`run_binary` are built on.
    """
    binary = resolve_binary()
    child_env = dict(os.environ)
    if env:
        child_env.update(env)

    input_bytes: bytes | None
    if stdin is None:
        input_bytes = None
    elif isinstance(stdin, bytes):
        input_bytes = stdin
    else:
        input_bytes = stdin.encode("utf-8")

    child = subprocess.Popen(
        [binary, *_with_globals(args, log_level)],
        env=child_env,
        stdin=subprocess.PIPE if input_bytes is not None else None,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )
    stdout = bytearray()
    stderr = bytearray()

    def drain(pipe: object, captured: bytearray, callback: Callable[[bytes], None] | None) -> None:
        while chunk := pipe.read(8192):  # type: ignore[attr-defined]
            captured.extend(chunk)
            if callback:
                callback(chunk)

    out_thread = threading.Thread(target=drain, args=(child.stdout, stdout, on_stdout))
    err_thread = threading.Thread(target=drain, args=(child.stderr, stderr, on_stderr))
    out_thread.start()
    err_thread.start()
    if input_bytes is not None and child.stdin is not None:
        child.stdin.write(input_bytes)
        child.stdin.close()
    exit_code = child.wait()
    out_thread.join()
    err_thread.join()
    return bytes(stdout), bytes(stderr), exit_code


def run(
    args: list[str],
    *,
    env: Mapping[str, str] | None = None,
    stdin: str | bytes | None = None,
    log_level: int | None = None,
    on_stdout: Callable[[bytes], None] | None = None,
    on_stderr: Callable[[bytes], None] | None = None,
) -> RawResult:
    """Run ``bsdkrun <args>`` to completion, buffering stdout/stderr.

    ``env`` is merged onto the current process environment. ``stdin``, if given,
    is piped to the child. ``log_level`` sets bsdkrun's global ``--log-level``
    (defaults to 0 — quiet).
    """
    out, err, code = _run_raw(
        args, env=env, stdin=stdin, log_level=log_level, on_stdout=on_stdout, on_stderr=on_stderr
    )
    return RawResult(
        stdout=out.decode("utf-8", "replace"),
        stderr=err.decode("utf-8", "replace"),
        exit_code=code,
    )


def run_binary(
    args: list[str],
    *,
    env: Mapping[str, str] | None = None,
    stdin: str | bytes | None = None,
    log_level: int | None = None,
) -> BinaryResult:
    """Run ``bsdkrun <args>`` and keep stdout as raw bytes.

    :func:`run` decodes stdout as UTF-8 with ``errors="replace"``, which quietly
    rewrites every invalid byte — harmless for JSON, ruinous for
    ``cp ID:path -`` reading an image. Only stderr is decoded here, because it
    is always a message.
    """
    out, err, code = _run_raw(args, env=env, stdin=stdin, log_level=log_level)
    return BinaryResult(stdout=out, stderr=err.decode("utf-8", "replace"), exit_code=code)


def run_checked(
    args: list[str],
    label: str,
    *,
    env: Mapping[str, str] | None = None,
    stdin: str | bytes | None = None,
    log_level: int | None = None,
    on_stdout: Callable[[bytes], None] | None = None,
    on_stderr: Callable[[bytes], None] | None = None,
) -> RawResult:
    """Like :func:`run`, but raise :class:`CommandFailed` on a non-zero exit."""
    result = run(
        args,
        env=env,
        stdin=stdin,
        log_level=log_level,
        on_stdout=on_stdout,
        on_stderr=on_stderr,
    )
    if result.exit_code != 0:
        raise CommandFailed(result.exit_code, result.stdout, result.stderr, label)
    return result


def spawn(
    args: list[str],
    *,
    env: Mapping[str, str] | None = None,
    log_level: int | None = None,
) -> int:
    """Run ``bsdkrun <args>`` inheriting the parent's stdio (interactive).

    Blocks until the child exits and returns its exit code. Used by
    :meth:`~bsdkrun.sandbox.Sandbox.shell`.
    """
    binary = resolve_binary()
    child_env = dict(os.environ)
    if env:
        child_env.update(env)
    completed = subprocess.run(
        [binary, *_with_globals(args, log_level)],
        env=child_env,
        check=False,
    )
    return completed.returncode

File: sdk/scala/src/main/scala/bsdkrun/Process.scala
package bsdkrun

import java.io.ByteArrayOutputStream
import java.nio.charset.StandardCharsets.UTF_8
import scala.jdk.CollectionConverters.*

/** Runs the `bsdkrun` CLI and captures its output.
  *
  * Every invocation is prefixed with the global `--log-level` flag (default 0)
  * so the SDK's captured output stays clean. Raise it for boot diagnostics.
  */
object Proc:

  /** The buffered result of a `bsdkrun` invocation. */
  final case class RawResult(stdout: String, stderr: String, exitCode: Int)

  /** Like [[RawResult]], but stdout is left as bytes.
    *
    * Decoding stdout as UTF-8 replaces every invalid byte — fine for JSON,
    * ruinous for `cp ID:path -` reading a PNG. Only stderr is decoded here,
    * because it is always a message.
    */
  final case class BinaryResult(stdout: Array[Byte], stderr: String, exitCode: Int)

  /** Options for one invocation. */
  final case class Options(
      env: Map[String, String] = Map.empty,
      stdin: Option[Array[Byte]] = None,
      logLevel: Int = 0,
      onStdout: Option[Array[Byte] => Unit] = None,
      onStderr: Option[Array[Byte] => Unit] = None
  )

  private def withGlobals(args: Seq[String], logLevel: Int): Seq[String] =
    Seq("--log-level", logLevel.toString) ++ args

  /** Run `bsdkrun <args>` and return `(stdout, stderr, exit code)` as bytes —
    * the undecoded form both [[run]] and [[runBinary]] are built on.
    */
  private def runRaw(
      args: Seq[String],
      opts: Options
  ): Either[BsdkrunError, (Array[Byte], Array[Byte], Int)] =
    Binary.resolve().flatMap: binary =>
      try
        val builder = new java.lang.ProcessBuilder((binary +: withGlobals(args, opts.logLevel)).asJava)
        opts.env.foreach((k, v) => builder.environment().nn.put(k, v))
        val proc = builder.start().nn

        // Feed stdin and drain both streams concurrently. A child that starts
        // producing output before it has consumed all of stdin would otherwise
        // deadlock against a sequential write — the same job Python's
        // `communicate()` does with its worker threads.
        val stdinThread = new Thread(() =>
          val os = proc.getOutputStream.nn
          try opts.stdin.foreach(os.write)
          catch case _: java.io.IOException => () // the child exited early
          finally os.close()
        )
        val stdout = new ByteArrayOutputStream()
        val stderr = new ByteArrayOutputStream()
        val outThread = drain(proc.getInputStream.nn, stdout, opts.onStdout)
        val errThread = drain(proc.getErrorStream.nn, stderr, opts.onStderr)

        stdinThread.start()
        outThread.start()
        errThread.start()
        val code = proc.waitFor()
        stdinThread.join()
        outThread.join()
        errThread.join()

        Right((stdout.toByteArray.nn, stderr.toByteArray.nn, code))
      catch
        case e: java.io.IOException =>
          Left(BsdkrunError.CommandFailed(-1, "", s"could not run $binary: ${e.getMessage}", args.mkString(" ")))

  private def drain(
      in: java.io.InputStream,
      into: ByteArrayOutputStream,
      callback: Option[Array[Byte] => Unit]
  ): Thread =
    new Thread(() =>
      val buf = new Array[Byte](8192)
      var n = in.read(buf)
      while n > 0 do
        into.write(buf, 0, n)
        callback.foreach(cb => cb(buf.take(n)))
        n = in.read(buf)
    )

  /** Run `bsdkrun <args>` to completion, buffering stdout/stderr as text. */
  def run(args: Seq[String], opts: Options = Options()): Either[BsdkrunError, RawResult] =
    runRaw(args, opts).map: (out, err, code) =>
      RawResult(new String(out, UTF_8), new String(err, UTF_8), code)

  /** Run `bsdkrun <args>` and keep stdout as raw bytes. */
  def runBinary(args: Seq[String], opts: Options = Options()): Either[BsdkrunError, BinaryResult] =
    runRaw(args, opts).map: (out, err, code) =>
      BinaryResult(out, new String(err, UTF_8), code)

  /** Like [[run]], but a non-zero exit becomes [[BsdkrunError.CommandFailed]]
    * tagged with `label`.
    */
  def runChecked(
      args: Seq[String],
      label: String,
      opts: Options = Options()
  ): Either[BsdkrunError, RawResult] =
    run(args, opts).flatMap: res =>
      if res.exitCode == 0 then Right(res)
      else Left(BsdkrunError.CommandFailed(res.exitCode, res.stdout, res.stderr, label))

  /** Run `bsdkrun <args>` inheriting this process's stdio, for the interactive
    * subcommands. Blocks until the child exits and returns its exit code.
    */
  def spawn(args: Seq[String], logLevel: Int = 0): Either[BsdkrunError, Int] =
    Binary.resolve().map: binary =>
      new java.lang.ProcessBuilder((binary +: withGlobals(args, logLevel)).asJava)
        .inheritIO()
        .nn
        .start()
        .nn
        .waitFor()
